packages/react-dom/src/client/ReactDOM.js

```diff
@@ -946,18 +946,28 @@ const DOMRenderer = ReactFiberReconciler({
       parentContainer: Container,
       type: string,
       props: Props,
+      index: number,
+      isReplaced: boolean,
```

Comment:
@acdlite Hi,

This PR is about showing a nice DOM diff and a component stack in the hydration mismatch warnings for developers.

To implement the diff I had to add two additional arguments 1) `index: number` and 2) `isReplaced: boolean` to the `ReactFiberReconciler`'s `didNotFindHydratable*` methods to pass down to the warning-generating functions 1) the index of the child node that's being hydrated within the parent node's hydratable children and 2) the fact that it's being replaced or inserted.

The `index` is required to show the minus/plus diff prefix at the correct line when looping through the parent node's hydratable children while constructing the diff.

The `isReplaced` fact is required to not add the minus diff prefix when the preceding sibling node was not in fact removed during hydration.

Example warning when the `<p>` element did not match and was replaced with the `<div>` element (`isReplaced: true`):
```
Warning: Expected server HTML to contain a matching <div>{['children ', …]}</div>
 in <div>nested<!-- --> <p>children <b>text</b></p></div>.
  <div>
    {'nested'}
    {' '}
-   <p>children <b>text</b></p>
+   <div>{['children ', …]}</div>
  </div>
    in div (at **)
    in Component (at **)
```

Example warning when the `<p>` element did match thus not removed/replaced, the `<div>` element was added (`isReplaced: false`):
```
Warning: Expected server HTML to contain a matching <div>{['children ', …]}</div>
 in <div>nested<!-- --> <p>children <b>text</b></p></div>.
  <div>
    {'nested'}
    {' '}
    <p>children <b>text</b></p>
+   <div>{['children ', …]}</div>
  </div>
    in div (at **)
    in Component (at **)
```

Could you please approve this change or recommend another direction?

Thank you.